[C10D] Support group_dst/group_src in c10d send/recv

wconstab opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #140827
* #140820
* #140815
* __->__ #140460

Partly addressing RFC 0042 (https://github.com/pytorch/rfcs/pull/71)

It's annoying that 'dst' (for send) ust be a global rank even when a
group is passed in.  But we can't easily change 'dst' without breaking
existing cases.

Furthermore, requiring use of 'global' dst breaks the less common usage
pattern of creating a new ProcessGroup object that is not connected to
the 'default group' and thus has no logical 'global' ranks.

cc @H-Huang @awgu @kwen2501 @wanchaol @fegin @fduwjj @wz337 @d4l3k @c-p-i-o

wconstab committed:
[C10D] Support group_dst/group_src in c10d collectives

Addressing RFC 0042 (https://github.com/pytorch/rfcs/pull/71)

It's annoying that 'dst' (for send) ust be a global rank even when a
group is passed in.  But we can't easily change 'dst' without breaking
existing cases.

[ghstack-poisoned]

wconstab commented on torch/distributed/distributed_c10d.py:
```diff
@@ -2322,8 +2322,20 @@ def send(
+        # TODO if we want to encourage migration to 'group_dst' arg and deprecate dst arg, should we support using 'group_dst'
+        # even when 'group' is None (for default group)?
+        assert group is not None, "Must specify group if using group_dst"
+        dst = get_global_rank(group, group_dst)
```
oops, this is dumb. i just realized that i can't implement group_dst as a function of global rank, since part of the reason for this PR is that 'group' might not be associated with the default-group at all, so get_global_rank function may not work.  We will need to invert the logic below so that 'group_dst' is the thing that we directly pass into the group api, and we convert dst into group_dst at the top if dst is passed in.  this will make the change a bit more invasive but still possible.

shuqiangzhang replied:
There is a global to group map mapping: _world.pg_group_ranks

wconstab replied:
yea, i realize that but the point is we want to not rely on any global stuff when using group_ argument.  This is to enable use cases where someone creates a ProcessGroup object directly and it's not registered with us.

wconstab committed:
Update on "[C10D] Support group_dst/group_src in c10d collectives"


Addressing RFC 0042 (https://github.com/pytorch/rfcs/pull/71)

It's annoying that 'dst' (for send) ust be a global rank even when a
group is passed in.  But we can't easily change 'dst' without breaking
existing cases.

cc H-Huang awgu kwen2501 wanchaol fegin fduwjj wz337 d4l3k c-p-i-o

[ghstack-poisoned]

wconstab committed:
Update on "[C10D] Support group_dst/group_src in c10d collectives"


Addressing RFC 0042 (https://github.com/pytorch/rfcs/pull/71)

It's annoying that 'dst' (for send) ust be a global rank even when a
group is passed in.  But we can't easily change 'dst' without breaking
existing cases.

cc H-Huang awgu kwen2501 wanchaol fegin fduwjj wz337 d4l3k c-p-i-o

[ghstack-poisoned]

wconstab committed:
Update on "[C10D] Support group_dst/group_src in c10d collectives"


Addressing RFC 0042 (https://github.com/pytorch/rfcs/pull/71)

It's annoying that 'dst' (for send) ust be a global rank even when a
group is passed in.  But we can't easily change 'dst' without breaking
existing cases.

Furthermore, requiring use of 'global' dst breaks the less common usage
pattern of creating a new ProcessGroup object that is not connected to
the 'default group' and thus has no logical 'global' ranks.

[ghstack-poisoned]

d4l3k reviewed (approved):
code looks good to me

Can we add some unit tests on this new logic?

c-p-i-o commented on torch/distributed/distributed_c10d.py:
```diff
@@ -2380,23 +2409,17 @@ def recv(
-            group_src_rank = get_group_rank(pg, src)
-            pg.recv([tensor], group_src_rank, tag).wait()
-        return src
+        group_src = _canonicalize_group_rank(group, src, group_src)
```
Can we use `_check_not_self_rank(group, group_src, "source")` here?

wconstab replied:
hmm, no, we apparently can't.  It causes this test to fail 
` python test/distributed/test_distributed_spawn.py TestDistBackendWithSpawn.test_batch_isend_irecv_nccl`

The test looks kinda dumb, i'm not sure if it really even intends to send/recv to itself, but that's what it does.  And I think (guess?) that nccl supports this so we probably shouldn't start asserting against it now...

c-p-i-o commented on torch/distributed/distributed_c10d.py:
```diff
@@ -1111,6 +1111,37 @@ def _check_tensor_list(param, param_name) -> None:
+def _check_not_self_rank(group: ProcessGroup, rank: int, rank_type: str):
+    if group.rank() == rank:
+        raise ValueError(
+            f"Invalid {rank_type} rank: {rank_type} rank should not be the same as "
```
Did you mean to print `rank_type` twice here?

wconstab replied:
Yea it's repetitive but I was preserving the original log message

fduwjj commented:
lg! maybe n00b question, what could be a use case for this?

kwen2501 commented on torch/distributed/distributed_c10d.py:
```diff
@@ -1111,6 +1111,37 @@ def _check_tensor_list(param, param_name) -> None:
+    Helper method to take _either_ a global rank or a group rank and produce a group rank.
+    """
+
+    if group_rank is not None:
+        assert global_rank is None, "Can't specify both group_rank and global_rank"
+    else:
+        assert global_rank is not None, "Must specify global_rank or group_rank"
```
nit: raise ValueError instead? assert can be optimized away with `-O` flag.

kwen2501 reviewed (approved):
LGTM. Thanks a lot.
Do you mind extending the change to `isend`, `irecv`, `broadcast` and `reduce` too?
Can be in a next PR.
Thanks!

wconstab commented:
> lg! maybe n00b question, what could be a use case for this?

Well, first I wanted this when developing pipelining. We often want to send to our next pipeline rank using the pp group but we are forced to convert our pp rank to global rank which is inconvenient.

Second the RFC0042 issue came up and using local ranks is important if you don't want to assume all pgs are part of the same world group.


@kwen2501 yea planning to add other ops in additional PRs.

wconstab committed:
Update on "[C10D] Support group_dst/group_src in c10d collectives"


Addressing RFC 0042 (https://github.com/pytorch/rfcs/pull/71)

It's annoying that 'dst' (for send) ust be a global rank even when a
group is passed in.  But we can't easily change 'dst' without breaking
existing cases.

Furthermore, requiring use of 'global' dst breaks the less common usage
pattern of creating a new ProcessGroup object that is not connected to
the 'default group' and thus has no logical 'global' ranks.

[ghstack-poisoned]

wconstab committed:
Update on "[C10D] Support group_dst/group_src in c10d send/recv"


Partly addressing RFC 0042 (https://github.com/pytorch/rfcs/pull/71)

It's annoying that 'dst' (for send) ust be a global rank even when a
group is passed in.  But we can't easily change 'dst' without breaking
existing cases.

Furthermore, requiring use of 'global' dst breaks the less common usage
pattern of creating a new ProcessGroup object that is not connected to
the 'default group' and thus has no logical 'global' ranks.

[ghstack-poisoned]

wconstab committed:
Update on "[C10D] Support group_dst/group_src in c10d send/recv"


Partly addressing RFC 0042 (https://github.com/pytorch/rfcs/pull/71)

It's annoying that 'dst' (for send) ust be a global rank even when a
group is passed in.  But we can't easily change 'dst' without breaking
existing cases.

Furthermore, requiring use of 'global' dst breaks the less common usage
pattern of creating a new ProcessGroup object that is not connected to
the 'default group' and thus has no logical 'global' ranks.

[ghstack-poisoned]

kwen2501 commented:
Sorry I have more ops to suggggggest:
`broadcast_object_list`, `scatter_object_list`, `send_object_list`, `recv_object_list`.

Just a heads-up: in the implementation of these object coll functions, we should use the `group_src` arg for the inner calls. Otherwise there would be a double conversion :)

wconstab committed:
Update on "[C10D] Support group_dst/group_src in c10d send/recv"


Partly addressing RFC 0042 (https://github.com/pytorch/rfcs/pull/71)

It's annoying that 'dst' (for send) ust be a global rank even when a
group is passed in.  But we can't easily change 'dst' without breaking
existing cases.

Furthermore, requiring use of 'global' dst breaks the less common usage
pattern of creating a new ProcessGroup object that is not connected to
the 'default group' and thus has no logical 'global' ranks.

[ghstack-poisoned]

wconstab committed:
Update on "[C10D] Support group_dst/group_src in c10d send/recv"


Partly addressing RFC 0042 (https://github.com/pytorch/rfcs/pull/71)

It's annoying that 'dst' (for send) ust be a global rank even when a
group is passed in.  But we can't easily change 'dst' without breaking
existing cases.

Furthermore, requiring use of 'global' dst breaks the less common usage
pattern of creating a new ProcessGroup object that is not connected to
the 'default group' and thus has no logical 'global' ranks.

[ghstack-poisoned]
